### ISSUE TITLE
Reusing parser produces unexpected behavior in certain cases

### ISSUE THREAD

**Issue by majgis** [Leonidas-from-XIV/node-xml2js#388]
If you are reusing a parser, with parseString having previously been called, and you call parseString with `<?xml version="1.0" encoding="UTF-8"?>` the callback will not be called until after the next call to parseString, with the result from this second call:
```
> parseString = xml2js.Parser().parseString
[Function]
> parseString('<?xml version="1.0" encoding="UTF-8"?><config></config>', (err, result)=> console.log('result', result)) && null
result { config: '' }
null
> parseString('<?xml version="1.0" encoding="UTF-8"?><config></config>', (err, result)=> console.log('result', result)) && null
result { config: '' }
null
> parseString('<?xml version="1.0" encoding="UTF-8"?><config></config>', (err, result)=> console.log('result', result)) && null
result { config: '' }
null
> parseString('<?xml version="1.0" encoding="UTF-8"?>', (err, result)=> console.log('result', result)) && null
null
> parseString('<?xml version="1.0" encoding="UTF-8"?><config><parameter></parameter></config>', (err, result)=> console.log('result', result)) && null
result { config: { parameter: [ '' ] } }
result { config: { parameter: [ '' ] } }
null
```

If you do not reuse a parser, this issue is avoided:
```
> parseString = xml2js.Parser().parseString
[Function]
> parseString('<?xml version="1.0" encoding="UTF-8"?>', (err, result)=> console.log('result', result)) && null
result null
null
```

**Comment by Leonidas-from-XIV** [comment]
I specifically warn against reusing the parser in the `README`.

**Comment by basemath** [comment]
This issue bit me pretty bad. The Readme does say you should make separate Parser instances, but ... dang. It seems like the Parser could at least throw an error more indicative of the actual problem, instead of saying there are syntax issues where there are none.

**Comment by majgis** [comment]
I would agree that creating a landmine and only a comment in the readme well below usage examples to save you isn't the greatest contract.  I would suggest that if calling reset() fixes the issue, it should be called  internally after each call to parseString.

**Comment by Leonidas-from-XIV** [comment]
Yes I am sorry. This is due to the way the API allows feeding in XML piece by piece which I would not recommend to do anymore.

There is unfortunately not much I can change about it without breaking the library for everybody else who is relying on being able to reuse the parser.

**Comment by Leonidas-from-XIV** [comment]
This is not a created landmine, this is one inherited from a very old version of xml2js that I took over. Over the course of time I added APIs that allow less accidental misuse, but didn't change the old behaviour because people might still depend on the old, wonky behaviour.

E.g. I reverted one completely feature in the latest release because people were depending on completely incorrect behaviour where an empty string is considered a valid XML file, so I am quite cautious of breaking things for more people.

Also, `reset()` is just a best effort thing. I really advise people not to reuse the parser but when they cannot, it's better than nothing.

I also plan to do a version of xml2js which gets rid of all this unclear behaviour, but I only have so much time after work.
